SCAutolib/models/gui.py

```diff
@@ -0,0 +1,304 @@
+from SCAutolib import run, logger
+
+import os
+from time import sleep, time
+import cv2
+import numpy as np
+import pandas as pd
+from io import StringIO
+import pytesseract
+import keyboard
+import uinput
+import functools
+
+
+class Screen():
+    """Captures the screenshots."""
+
+    def __init__(self, directory: str):
+        self.directory = directory
+        self.screenshot_num = 1
+
+    def screenshot(self, timeout: float = 30):
+        """Runs ffmpeg to take a screenshot.
+
+        :return: filename of the screenshot
+        :rtype: str
+        """
+
+        logger.debug(f"Taking screenshot number {self.screenshot_num}")
+
+        filename = f'{self.directory}/{self.screenshot_num}.png'
+        t_end = time() + timeout
+        captured = False
+
+        while time() < t_end and not captured:
+            out = run(['ffmpeg', '-hide_banner', '-y', '-f',
+                       'kmsgrab', '-i', '-', '-vf', 'hwdownload,format=bgr0',
+                       '-frames', '1', '-update', '1',
+                       filename], check=False, print_=False)
+
+            if out.returncode == 0:
+                captured = True
+
+        if not captured:
+            raise Exception('Could not capture screenshot within timeout.')
+
+        self.screenshot_num += 1
+        return filename
+
+
+class Mouse():
+    def __init__(self):
+        run(['modprobe', 'uinput'], check=True)
+
+        # Maximum coordinate for both axis
+        self.ABS_MAX = 2**16
+
+        # initialize the uinput device
+        self.device = uinput.Device((
+            uinput.ABS_X + (0, self.ABS_MAX, 0, 0),
+            uinput.ABS_Y + (0, self.ABS_MAX, 0, 0),
+            uinput.BTN_LEFT,
+            uinput.BTN_MIDDLE,
+            uinput.BTN_RIGHT,
+            uinput.REL_WHEEL,
+        ))
+
+        self.CLICK_HOLD_TIME = 0.1
+
+    def move(self, x: float, y: float):
+        """Moves the mouse cursor to specified absolute coordinate."""
+
+        logger.info(f'Moving mouse to {x, y})')
+
+        for uinput_axis, value in [(uinput.ABS_X, x), (uinput.ABS_Y, y)]:
+            # Check if value between 0 and 1
+            assert ((value >= 0) and (value <= 1))
+            converted = int(value * self.ABS_MAX)
+            self.device.emit(uinput_axis, converted, syn=False)
+
+        # Both axis move at once
+        self.device.syn()
+
+    def click(self, button: str = 'left'):
+        """Clicks the any button of the mouse.
```

Comment:
type information missing in this docstring. Better phrasing would be "clicks selected button"